In [ ]:
# import modules
%matplotlib inline
import warnings
warnings.filterwarnings("ignore")

import datetime
import pandas as pd
import numpy as np
import string
import matplotlib.pyplot as plt
from wordcloud import WordCloud, STOPWORDS
import seaborn as sns
import math
import nltk
from collections import Counter
import os
import shutil
from tqdm.notebook import tqdm
tqdm.pandas()
import re
from nltk.corpus import stopwords
from nltk.sentiment.vader import SentimentIntensityAnalyzer
from tqdm import tqdm
tqdm.pandas()
import pickle
from sklearn.preprocessing import MinMaxScaler
import csv
from scipy import sparse
from sklearn.preprocessing import MinMaxScaler

In [ ]:
from sklearn.model_selection import RandomizedSearchCV
from sklearn.metrics import mean_squared_log_error as rmsle
from sklearn.linear_model import Lasso
from sklearn.linear_model import SGDRegressor
from sklearn.linear_model import Ridge
from sklearn.ensemble import RandomForestRegressor
from sklearn.utils.fixes import loguniform
from multiprocessing import  Pool
from sklearn.metrics import mean_squared_error

# importing the .py file which contains the dataCleaning, dataProcessing and converting to vectors
import preprocess_data_clean as pde  

[nltk_data] Downloading package stopwords to
[nltk_data]     /home/rahul_rbbisht1050/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package vader_lexicon to
[nltk_data]     /home/rahul_rbbisht1050/nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


In [ ]:
# !wget --header="Host: storage.googleapis.com" --header="User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.77 Safari/537.36" --header="Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9" --header="Accept-Language: en-US,en;q=0.9" --header="Referer: https://www.kaggle.com/" "https://storage.googleapis.com/kagglesdsdata/competitions/7559/44327/test.tsv.7z?GoogleAccessId=web-data@kaggle-161607.iam.gserviceaccount.com&Expires=1625901261&Signature=nQ81vqsNNxCnfF2jjUxcwQfN3zPAla934vi2qcjyZtlxWDJRMpuI9Bojlwj1Yc8qwLIXbBfCxLOsPbvjETMO3ed1EMI%2BuueARI%2FcYW5CZ0BaD58GPt1uSkOMNtc2aZXWNIyE%2FzMFuffDE2ZrvBA69zeoZWFO%2BcWZNUsLMjAEl8Iym%2BENUqO8dPx%2BodJNSNmQHXHXetAxmPf%2BII7RxRP7%2BxsL4fDxRFflPeBzbhF%2BayxjtmgznLuh8VTngEewfuWfMs9YgAXE8WgQqC20q4S3%2F2xgHr9NZb4g2CWthC6W4xsOgOIKWviW9g6XutCcvZCQTFcT4ZmIyJEC1H0zgJUTXQ%3D%3D&response-content-disposition=attachment%3B+filename%3Dtest.tsv.7z" -c -O 'test.tsv.7z'

In [ ]:
# !wget --header="Host: storage.googleapis.com" --header="User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.77 Safari/537.36" --header="Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9" --header="Accept-Language: en-US,en;q=0.9" --header="Referer: https://www.kaggle.com/" "https://storage.googleapis.com/kagglesdsdata/competitions/7559/44327/test_stg2.tsv.zip?GoogleAccessId=web-data@kaggle-161607.iam.gserviceaccount.com&Expires=1625896403&Signature=qPoxBWeRrpSeWC7nE5E%2BNmpKyfzVNQty3XFEzSqbOUHuXUHXaMFLpoRWAJrqtr3wHujYe9%2B310GfoPy6tbfgUuHIjV6hvUYTG5dI%2BgeTtLmuumCFxKNAhvnd3hqbT7lCggv57iZQbmjS7LesPhLHipVKqSX5%2FTD9ljV4lr5mFxOrPSeqFKtcW%2B8Oq31qdb23USXWJ8CmYPJEgjckGXRqB3X%2Bsunzb571r7%2FL4rM5bYKB7al0u%2FQf9OEm219OKPKesPF9TpafPEKB3pyTdzqx9GFLkl7%2F3Bp1LNIyYXMDX8oGJhWcHgvxYENT5NThoKdC6GNLUA8by2cnbLtLL5mdJw%3D%3D&response-content-disposition=attachment%3B+filename%3Dtest_stg2.tsv.zip" -c -O 'test_stg2.tsv.zip'

--2021-07-07 07:04:49--  https://storage.googleapis.com/kagglesdsdata/competitions/7559/44327/test_stg2.tsv.zip?GoogleAccessId=web-data@kaggle-161607.iam.gserviceaccount.com&Expires=1625896403&Signature=qPoxBWeRrpSeWC7nE5E%2BNmpKyfzVNQty3XFEzSqbOUHuXUHXaMFLpoRWAJrqtr3wHujYe9%2B310GfoPy6tbfgUuHIjV6hvUYTG5dI%2BgeTtLmuumCFxKNAhvnd3hqbT7lCggv57iZQbmjS7LesPhLHipVKqSX5%2FTD9ljV4lr5mFxOrPSeqFKtcW%2B8Oq31qdb23USXWJ8CmYPJEgjckGXRqB3X%2Bsunzb571r7%2FL4rM5bYKB7al0u%2FQf9OEm219OKPKesPF9TpafPEKB3pyTdzqx9GFLkl7%2F3Bp1LNIyYXMDX8oGJhWcHgvxYENT5NThoKdC6GNLUA8by2cnbLtLL5mdJw%3D%3D&response-content-disposition=attachment%3B+filename%3Dtest_stg2.tsv.zip
Resolving storage.googleapis.com (storage.googleapis.com)... 216.58.196.208, 216.58.221.48, 172.217.160.240, ...
Connecting to storage.googleapis.com (storage.googleapis.com)|216.58.196.208|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 308669128 (294M) [application/zip]
Saving to: ‘test_stg2.tsv.zip’

test_stg2.tsv.zip   100%[==

In [ ]:
!unzip test_stg2.tsv.zip

Archive:  test_stg2.tsv.zip
  inflating: test_stg2.tsv           


In [ ]:
# test data for submission

test2 = pd.read_csv('test_data/test_stg2.tsv', sep='\t')
print(test2.shape)
test2.head()

(3460725, 7)


test_id                                      name  item_condition_id  \
0        0  Breast cancer "I fight like a girl" ring                  1   
1        1  25 pcs NEW 7.5"x12" Kraft Bubble Mailers                  1   
2        2                                 Coach bag                  1   
3        3                             Floral Kimono                  2   
4        4                          Life after Death                  3   

                                    category_name brand_name  shipping  \
0                             Women/Jewelry/Rings        NaN         1   
1         Other/Office supplies/Shipping Supplies        NaN         1   
2  Vintage & Collectibles/Bags and Purses/Handbag      Coach         1   
3                         Women/Sweaters/Cardigan        NaN         0   
4             Other/Books/Religion & Spirituality        NaN         1   

                                    item_description  
0                                             Size 7  
1  25 pcs NEW 7.5"x12" Kraft Bubble Mailers Lined...  
2  Brand new coach bag. Bought for [rm] at a Coac...  
3  -floral kimono -never worn -lightweight and pe...  
4  Rediscovering life after the loss of a loved o...

In [ ]:
# getting the dtypes of columns present in feature
test2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3460725 entries, 0 to 3460724
Data columns (total 7 columns):
 #   Column             Dtype 
---  ------             ----- 
 0   test_id            int64 
 1   name               object
 2   item_condition_id  int64 
 3   category_name      object
 4   brand_name         object
 5   shipping           int64 
 6   item_description   object
dtypes: int64(3), object(4)
memory usage: 184.8+ MB


# Preprocessing the test data

preprocessDataCleaningFeatureEngineering imported as pde contains a class which has 3 fuctions -<br>
preprocess()  - for preprocessing the data <br>
calculate_unique_brands() - for finding the unique brands. This is useful for filling the brands value. <br>
feature_engineering() - function for feature engineering<br>
vectorization() - function to convert features into vectors. <br>

<br>



In [ ]:
# https://towardsdatascience.com/make-your-own-super-pandas-using-multiproc-1c04f41944a1
    
def parallelize_dataframe(df, func, n_cores=8):
    '''
        function to split the dataframe in to chunks and parallelizing those chunks based on the number of cores present.
        input: dataframe, function_to_be applied on chunks, n_cores: number of cores
        output: combined dataframe from different chunks
    '''
    df_split = np.array_split(df, n_cores)
    pool = Pool(n_cores)
    df = pd.concat(pool.map(func, df_split))
    pool.close()
    pool.join()
    return df

In [ ]:
# calling the class present inside the preprocessDataCleaningFeatureEngineering.py file
pde_class = pde.DataCleaning_Preprocessing()

In [ ]:
# calling the preprocess function present inside the class and parallelizing the dataframe
test2_preprocess = parallelize_dataframe(test2, pde_class.preprocess)

In [ ]:
# function to calculate the unique brands present
pde_class.calculate_unique_brands(test2_preprocess)

In [ ]:
# calling the feature_engineering function present inside the class
test2_feature_engineering = parallelize_dataframe(test2_preprocess, pde_class.feature_engineering)

The shape of test data set is (3460725, 7). It took about 1 hour 17 minutes to run the above cell parallelly on a 8 core cpu in 
GCP (google cloud platform).

In [ ]:
# test2_feature_engineering.to_csv('mercari/test2_feature_engineering.csv')

There are some of the datapoints where the text feature are written in japanese language. After preprocessing the text feature
it returns the empty string for the respective datapoints.

In [ ]:
# filling the Nan datapoints which we got as a result of preprocessing
test2_feature_engineering.name.fillna('unk_name', inplace=True)
test2_feature_engineering.item_description.fillna('unk_desc', inplace=True)

In [ ]:
# reading the test csv file
test2_feature_engineering = pd.read_csv('mercari/test2_feature_engineering.csv')

In [ ]:
def vectorization(data):

    '''
        function to convert the data into its respective vector form
        input - dataframe
        output - csr_matrix with horizontal stacking of features
    '''
    print('\n')
    start_time = datetime.datetime.now()

    print('*'*20 + 'Converting the features into vectors...' + '*' * 20)
    print()

    print(data.columns)

    import os.path
    if not os.path.isfile('glove.6B.100d.txt'):
        print('please download glove vector first!')
        return  

    print('loading the glove vector...')
    words_dict = dict()
    f = open('glove.6B.100d.txt')
    for line in f:
        values = line.split()
        word = values[0]
        coefs = np.asarray(values[1:], dtype='float32')
        words_dict[word] = coefs
    f.close()
    print('glove vector loaded!')


    def sentence_word2vec(sentence):
        '''
            function to convert text_feature to respective vector form using glove_vector
        '''
        vector = np.zeros(100)
        for word in sentence.split():
            if word in words_dict:
                vector += words_dict[word]

        return vector


    print('converting the concatenation of name_description column to its respective vector form...')
    name_desc_vector = data.name_desc.astype(str).progress_apply(lambda x: sentence_word2vec(x))
    print('\nconversion of name_desc to vector completed!')
    print('-' * 100)
    print()

    print('converting the concatenation of name, brand, sub_categories column to its respective vector form...')
    name_brand_cat_vector = data.name_brand_cat.astype(str).progress_apply(lambda x: sentence_word2vec(x))
    print('\nconversion of name_brand_cat to vector completed!')
    print('-' * 100)
    print()

    data.drop(columns=['name_desc', 'name_brand_cat'], inplace=True)

    print('converting text_feature to vector form...')
    name_desc_vector = sparse.csc_matrix(name_desc_vector.values.tolist())
    name_brand_cat_vector = sparse.csc_matrix(name_brand_cat_vector.values.tolist())
    print('converted text features to vector form')
    print('-'*100)
    print()

    cheap_brand = pickle.load(open('cheap_brand_set', "rb"))
    affordable_brand = pickle.load(open('affordable_brand_set', "rb"))
    expensive_brand = pickle.load(open('expensive_brand_set', "rb"))

    def fill_brand_category(brand_name):
        try:
            if brand_name in cheap_brand:
                return 'cheap'
            elif brand_name in affordable_brand:
                return 'affordable'
            elif brand_name in expensive_brand:
                return 'expensive'
            else:
                return 'affordable'
        except:
            return 'affordable'


    data['categorise_brand'] = data['brand_name'].apply(lambda x: fill_brand_category(x))


    def ordinal_encoder_load(feature_name, file_name):
        encoder = pickle.load(open('encoder/' + file_name + '_ordinal_encoder.pkl', 'rb'))
        encoder_values = encoder.transform(data[feature_name].astype(str).values.reshape(-1,1))

        imputer = pickle.load(open('imputer/' + file_name + '_imputer.pkl', 'rb'))
        imputer_values = imputer.transform(encoder_values)
        return imputer_values

    print('-'*100)
    print('applying label encoder and scaling the features...')


    train_brand_name = ordinal_encoder_load('brand_name', 'brand_name')
    train_category_brand = ordinal_encoder_load('categorise_brand', 'categorise_brand')
    train_category = ordinal_encoder_load('category_name', 'category_name')
    train_main_category = ordinal_encoder_load('main_category', 'main_category')
    train_sub_category1 = ordinal_encoder_load('brand_name', 'sub_category1')
    train_sub_category2 = ordinal_encoder_load('brand_name', 'sub_category2')

    # scaling the label encoded features
    brand_scaler = pickle.load(open('scaler/brand_scaler.pkl', 'rb'))
    train_brand_name =  brand_scaler.transform(train_brand_name)

    category_brand_scaler = pickle.load(open('scaler/category_brand_scaler.pkl', "rb"))
    train_category_brand  = category_brand_scaler.transform(train_category_brand)

    category_scaler = pickle.load(open('scaler/category_scaler.pkl', 'rb'))
    train_category =  category_scaler.transform(train_category)

    main_category_scaler = pickle.load(open('scaler/main_category_scaler.pkl', 'rb'))
    train_main_category =  main_category_scaler.transform(train_main_category)

    main_sub_category1_scaler = pickle.load(open('scaler/main_sub_category1_scaler.pkl', 'rb'))
    train_sub_category1 =  main_sub_category1_scaler.transform(train_sub_category1)

    main_sub_category2_scaler = pickle.load(open('scaler/main_sub_category2_scaler.pkl', 'rb'))
    train_sub_category2 =  main_sub_category2_scaler.transform(train_sub_category2)

    train_brand_name = sparse.csr_matrix(train_brand_name)
    train_category_brand = sparse.csr_matrix(train_category_brand)
    train_category = sparse.csr_matrix(train_category)
    train_main_category = sparse.csr_matrix(train_main_category)
    train_sub_category1 = sparse.csr_matrix(train_sub_category1)
    train_sub_category2 = sparse.csr_matrix(train_sub_category2)

    print('label encoding and scaling of features done!')
    print('-'*100)

    data.drop(columns=['brand_name','category_name', 'main_category', 'sub_category1', 'sub_category2'], inplace=True)

    data['item_condition_id'] = data['item_condition_id'] / 5.0

    len_name_scaler = pickle.load(open('scaler/len_name_scaler.pkl', 'rb'))
    data['len_name'] =  len_name_scaler.transform(data['len_name'].values.reshape(-1,1))

    len_item_description_scaler = pickle.load(open('scaler/len_item_description_scaler.pkl', 'rb'))
    data['len_item_description'] =  len_item_description_scaler.transform(data['len_item_description'].values.reshape(-1,1))

    print('feature scaling done!')



    data_hstack = sparse.hstack((data['item_condition_id'].values.reshape(-1,1),data['shipping'].values.reshape(-1,1),\
                                train_brand_name, train_category_brand, train_category, train_main_category,\
                                train_sub_category1, train_sub_category2, data['len_name'].values.reshape(-1,1),\
                                data['len_item_description'].values.reshape(-1,1), data['brand_value'].values.reshape(-1,1),\
                                data['pos'].values.reshape(-1,1),\
                                data['neg'].values.reshape(-1,1), data['neu'].values.reshape(-1,1), name_desc_vector, \
                                name_brand_cat_vector)).tocsr()

    print('time taken to execute the cell : ', datetime.datetime.now()- start_time)
    print()


    print('*'*30 + 'feature vectorization is done!' + '*' * 30)

    return data_hstack

In [ ]:
# calling the vectorization function present inside the class
text2_vectorization = vectorization(test2_feature_engineering)



********************Converting the features into vectors...********************

Index(['Unnamed: 0', 'Unnamed: 0.1', 'test_id', 'name', 'item_condition_id',
       'category_name', 'brand_name', 'shipping', 'item_description',
       'main_category', 'sub_category1', 'sub_category2', 'len_name',
       'len_item_description', 'name_desc', 'name_brand_cat', 'brand_value',
       'pos', 'neg', 'neu'],
      dtype='object')
loading the glove vector...
glove vector loaded!
converting the concatenation of name_description column to its respective vector form...


100%|█████████████████████████████████████████████████████████████████| 3460725/3460725 [01:39<00:00, 34770.73it/s]



conversion of name_desc to vector completed!
----------------------------------------------------------------------------------------------------

converting the concatenation of name, brand, sub_categories column to its respective vector form...


100%|█████████████████████████████████████████████████████████████████| 3460725/3460725 [01:07<00:00, 50950.68it/s]



conversion of name_brand_cat to vector completed!
----------------------------------------------------------------------------------------------------

converting text_feature to vector form...
converted text features to vector form
----------------------------------------------------------------------------------------------------

----------------------------------------------------------------------------------------------------
applying label encoder and scaling the features...
label encoding and scaling of features done!
----------------------------------------------------------------------------------------------------
feature scaling done!
time taken to execute the cell :  0:04:51.855130

******************************feature vectorization is done!******************************


In [ ]:
# sparse.save_npz('mercari/X_test2_hstack.npz', text2_vectorization)

In [ ]:
# loading the vectorizer
X_test2 =  sparse.load_npz('mercari/X_test2_hstack.npz')

# Baseline model

In [ ]:
data = pd.read_csv('mercari/train1.csv')

In [ ]:
price_by_brand_median = data.groupby('brand_name').median('price').price.to_dict()

In [ ]:
data['brand_price_median'] = data.brand_name.apply(lambda x: price_by_brand_median[x])

In [ ]:
rmsle(data.price, data['brand_price_median'])

0.43286602410041686

<br><br>

## Loading the sparse array

In [ ]:
# loading the train sparse array
X_train = sparse.load_npz('mercari/X_train_hstack.npz')
y_train = sparse.load_npz('mercari/y_train.npz').todense().tolist()

In [ ]:
# loading the test sparse array
X_test =  sparse.load_npz('mercari/X_test_hstack.npz' )
y_test = sparse.load_npz('mercari/y_test.npz').todense().tolist()

In [ ]:
y_train = np.log1p(y_train)
y_test  = np.log1p(y_test)

In [ ]:
# without normalizing the price the models performed very bad. 
# So after some research found that by converting the price to log prices and by normalizing them the model performed well.

# Normalizing the price
scaler_price = MinMaxScaler()
y_train_scaled = scaler_price.fit_transform(y_train)
y_test_sclaed =  scaler_price.transform(y_test)
pickle.dump(scaler_price, open('scaler/scaler_price.pkl','wb'))

In [ ]:
scaler_price = pickle.load(open('scaler/scaler_price.pkl', 'rb'))

In [ ]:
def hyperparameter_tuning(x, y, model, params, cv, model_name):
    '''
        The function is used for finding out the hyperparameters using RandomSearchCV.
        input: data_points(x,y) to which the model has to fit, 
                model: linear_model for which the parameters are to be found,
                params: parameters ,
                cv: cross validation which is of int type,
                model_name: model name of string type 

        returns: best_hyperparameters
    '''
    start_time = datetime.datetime.now()
    clf = RandomizedSearchCV(estimator = model, param_distributions = params,n_jobs=6, cv=cv, random_state=42, return_train_score=True)
    clf.fit(x,y)
    print('Calculating the parameters for {} model'.format(model_name))
    print()
    print('-'*40)
    print('best hyperparameters are:-',clf.best_params_)
    
    print()
    print('time taken - ',datetime.datetime.now() - start_time)

    return clf

In [ ]:
# using the root_mean_squared_error and not the rmsle because the prices are converted to log values
def root_mean_squared_error(y_true, y_true_pred):
    '''
        function to calculate the root mean squared error
    '''
    return np.sqrt(mean_squared_error(y_true, y_true_pred))

### SGDRegressor

In [ ]:
model = SGDRegressor()

params = dict()
params['alpha'] = loguniform(1e-7, 10)
best_params = hyperparameter_tuning(X_train, y_train_scaled, model, params, 3, 'SGDRegressor') 

/home/rahul_rbbisht1050/.local/lib/python3.7/site-packages/sklearn/utils/validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)
/home/rahul_rbbisht1050/.local/lib/python3.7/site-packages/sklearn/linear_model/_stochastic_gradient.py:1228: ConvergenceWarning: Maximum number of iteration reached before convergence. Consider increasing max_iter to improve the fit.
  ConvergenceWarning)
/home/rahul_rbbisht1050/.local/lib/python3.7/site-packages/sklearn/utils/validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)
/home/rahul_rbbisht1050/.local/lib/python3.7/site-packages/sklearn/linear_model/_stochastic_gradient.py:1228: ConvergenceWarning: Maximum number of iteration reached before convergence. Conside

/home/rahul_rbbisht1050/.local/lib/python3.7/site-packages/sklearn/utils/validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)
/home/rahul_rbbisht1050/.local/lib/python3.7/site-packages/sklearn/linear_model/_stochastic_gradient.py:1228: ConvergenceWarning: Maximum number of iteration reached before convergence. Consider increasing max_iter to improve the fit.
  ConvergenceWarning)
/home/rahul_rbbisht1050/.local/lib/python3.7/site-packages/sklearn/utils/validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)
/home/rahul_rbbisht1050/.local/lib/python3.7/site-packages/sklearn/linear_model/_stochastic_gradient.py:1228: ConvergenceWarning: Maximum number of iteration reached before convergence. Conside

Calculating the parameters for SGDRegressor model

----------------------------------------
best hyperparameters are:- {'alpha': 4.033800832600378}

time taken -  2:08:53.203130


In [ ]:
sgd_regressor_model = SGDRegressor(alpha=4.033800832600378)
sgd_regressor_model.fit(X_train, y_train_scaled)
y_train_pred = sgd_regressor_model.predict(X_train)
y_test_pred  = sgd_regressor_model.predict(X_test)


In [ ]:
y_train_pred = scaler_price.inverse_transform(y_train_pred.reshape(-1,1))
y_test_pred =  scaler_price.inverse_transform(y_test_pred.reshape(-1,1))

In [ ]:
print('Train error: ',root_mean_squared_error(y_train, y_train_pred))
print('Test  error: ',root_mean_squared_error(y_test,  y_test_pred))

Train error:  1.2693467728931624
Test  error:  1.2723025954460385


In [ ]:
# saving the model
pickle.dump(sgd_regressor_model, open('models/sgd_regressor_model.pkl', 'wb'))

In [ ]:
# loaidng the model
sgd_regressor = pickle.load(open('models/sgd_regressor_model.pkl', 'rb'))

pred1_test = sgd_regressor.predict(X_test2)

submission_sgd = pd.DataFrame(test2.test_id, columns=['test_id'])

submission_sgd['price'] = np.expm1(scaler_price.inverse_transform(pred1_test.reshape(-1,1)))

In [ ]:
submission_sgd.to_csv('submission/submission_sgd.csv')

### Lasso Regression

In [ ]:
model = Lasso()

params = dict()
params['alpha'] = [0.0000001, 0.00001, 0.001, 0.1, 1, 3, 5, 7, 9]
hyperparameters_ = hyperparameter_tuning(X_train, y_train_scaled, model, params, 3, 'Lasso Regressor')


/home/rahul_rbbisht1050/.local/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:516: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 633.2432479492218, tolerance: 1.2752720590094584
  max_iter, tol, rng, random, positive)
/home/rahul_rbbisht1050/.local/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:516: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 321.31156095767165, tolerance: 1.280136329427535
  max_iter, tol, rng, random, positive)
/home/rahul_rbbisht1050/.local/lib/python3.7/site-packages/sklearn/linear_model/_coordinate_descent.py:516: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 613.4679720904151, tolerance: 1.2790246086164758
  max_iter, tol, rng, random, positive)


Calculating the parameters for Lasso Regressor model

----------------------------------------
best hyperparameters are:- {'alpha': 1e-07}

time taken -  1:14:55.094241


In [ ]:
lasso_regressor_model = Lasso(alpha=1e-07)
lasso_regressor_model.fit(X_train,y_train_scaled)
y_train_pred = lasso_regressor_model.predict(X_train)
y_test_pred  = lasso_regressor_model.predict(X_test)


In [ ]:
y_train_pred = scaler_price.inverse_transform(y_train_pred.reshape(-1,1))
y_test_pred =  scaler_price.inverse_transform(y_test_pred.reshape(-1,1))

In [ ]:
print('Train error: ',root_mean_squared_error(y_train, y_train_pred))
print('Test  error: ',root_mean_squared_error(y_test,  y_test_pred))

Train error:  0.6494764360543506
Test  error:  0.6487308139438533


In [ ]:
pickle.dump(lasso_regressor_model, open('models/lasso_regressor_model.pkl', 'wb'))

In [ ]:
lasso_regressor = pickle.load(open('models/lasso_regressor_model.pkl', 'rb'))

pred2_test = lasso_regressor.predict(X_test2)
# pred2_test =  lasso_regressor.predict(X_test)

submission_lasso = pd.DataFrame(test2.test_id, columns=['test_id'])

submission_lasso['price'] = np.expm1(scaler_price.inverse_transform(pred2_test.reshape(-1,1)))

In [ ]:
submission_lasso.to_csv('submission/submission_lasso.csv')

### Ridge Regression

In [ ]:
model = Ridge(fit_intercept=False)

params = dict()
params['solver'] = ['cholesky','lsqr']              
params['alpha'] = loguniform(1e-7, 12)
best_params = hyperparameter_tuning(X_train, y_train_scaled, model, params, 3, 'Ridge_Regressor')


Calculating the parameters for Ridge_Regressor model

----------------------------------------
best hyperparameters are:- {'alpha': 0.00686547196906466, 'solver': 'cholesky'}

time taken -  0:08:07.860483


In [ ]:
ridge_regressor_model = Ridge(alpha=0.00686547196906466,solver='lsqr', fit_intercept=False)
ridge_regressor_model.fit(X_train,y_train_scaled)
y_train_pred = ridge_regressor_model.predict(X_train)
y_test_pred  = ridge_regressor_model.predict(X_test)


In [ ]:
y_train_pred = scaler_price.inverse_transform(y_train_pred.reshape(-1,1))
y_test_pred =  scaler_price.inverse_transform(y_test_pred.reshape(-1,1))

In [ ]:
print('Train error: ',root_mean_squared_error(y_train, y_train_pred))
print('Test  error: ',root_mean_squared_error(y_test,  y_test_pred))

Train error:  0.6622761382924385
Test  error:  0.6618423662958506


In [ ]:
pickle.dump(ridge_regressor_model, open('models/ridge_regressor_model.pkl', 'wb'))

In [ ]:
ridge_regressor = pickle.load(open('models/ridge_regressor_model.pkl', 'rb'))

pred3_test = ridge_regressor.predict(X_test2)
# pred3_test =  ridge_regressor.predict(X_test)

submission_ridge = pd.DataFrame(test2.test_id, columns=['test_id'])

submission_ridge['price'] = np.expm1(scaler_price.inverse_transform(pred3_test.reshape(-1,1)))

In [ ]:
submission_ridge.to_csv('submission/submission_ridge.csv')

### LGBM Regression

In [ ]:
from lightgbm import LGBMRegressor

In [ ]:
params = dict()
params={'learning_rate':[0.1,0.3,0.5,0.6],'max_depth':[5,8,12,15],'n_estimators':[50,100,150,200],'num_leaves':[15,25,50,75],'boosting_type':['gbdt']}
lgbm_params={'sub_sample':0.9,'colsample_bytree':0.8,'min_child_samples':50,'objective':'regression'}
model = LGBMRegressor(**lgbm_params)


best_params = hyperparameter_tuning(X_train, y_train_scaled, model, params, 3, 'LGBM Regressor')


[LightGBM] [Warning] Unknown parameter: sub_sample
Calculating the parameters for LGBM Regressor model

----------------------------------------
best hyperparameters are:- {'num_leaves': 75, 'n_estimators': 100, 'max_depth': 12, 'learning_rate': 0.5, 'boosting_type': 'gbdt'}

time taken -  0:17:21.283378


In [ ]:
params = {'num_leaves': 75, 'n_estimators': 100, 'max_depth': 12, 'learning_rate': 0.5, 'boosting_type': 'gbdt'}
lgbm_params={'sub_sample':0.9,'colsample_bytree':0.8,'min_child_samples':50,'objective':'regression'}
lgbm_regressor_model = LGBMRegressor(**params,**lgbm_params)
lgbm_regressor_model.fit(X_train,y_train_scaled)
y_train_pred = lgbm_regressor_model.predict(X_train)
y_test_pred  = lgbm_regressor_model.predict(X_test)


[LightGBM] [Warning] Unknown parameter: sub_sample


In [ ]:
y_train_pred = scaler_price.inverse_transform(y_train_pred.reshape(-1,1))
y_test_pred =  scaler_price.inverse_transform(y_test_pred.reshape(-1,1))

In [ ]:
print('Train error: ',root_mean_squared_error(y_train, y_train_pred))
print('Test  error: ',root_mean_squared_error(y_test,  y_test_pred))

Train error:  0.5081683785557494
Test  error:  0.5264128509090326


In [ ]:
pickle.dump(lgbm_regressor_model, open('models/lgbm_regressor_model.pkl', 'wb'))

In [ ]:
lgbm_regressor = pickle.load(open('models/lgbm_regressor_model.pkl', 'rb'))

pred4_test = lgbm_regressor.predict(X_test2)
# pred4_test =  lgbm_regressor.predict(X_test)

submission_lgbm = pd.DataFrame(test2.test_id, columns=['test_id'])

submission_lgbm['price'] = np.expm1(scaler_price.inverse_transform(pred4_test.reshape(-1,1)))

In [ ]:
submission_lgbm.to_csv('submission/submission_lgbm.csv')

In [ ]:

from prettytable import PrettyTable
  
# Specify the Column Names while initializing the Table
x = PrettyTable(["model", "train score", "test score", "submission score"])
  
# Add rows
x.add_row(["SGD Regressor", "1.26934", "1.27230", "1.21826"])
x.add_row(["Lasso Regressor", "0.64947", "0.64873", "2.57205", ])
x.add_row(["Ridge Regressor", "0.50816", "0.52641", "0.71395"])
x.add_row(["LGBM Regressor", "0.50816", "0.52641", "0.70646"])
print()
print()
print(x)
print()



+-----------------+-------------+------------+------------------+
|      model      | train score | test score | submission score |
+-----------------+-------------+------------+------------------+
|  SGD Regressor  |   1.26934   |  1.27230   |     1.21826      |
| Lasso Regressor |   0.64947   |  0.64873   |     2.57205      |
| Ridge Regressor |   0.50816   |  0.52641   |     0.71395      |
|  LGBM Regressor |   0.50816   |  0.52641   |     0.70646      |
+-----------------+-------------+------------+------------------+

